src/java.base/share/classes/jdk/internal/classfile/impl/BufWriterImpl.java

```diff
@@ -152,6 +157,55 @@ public void writeBytes(BufWriterImpl other) {
         writeBytes(other.elems, 0, other.offset);
     }
 
+    @SuppressWarnings("deprecation")
+    void writeUTF(String str) {
+        int strlen = str.length();
+
+        int countGreaterThanZero = 0;
+        byte coder = JLA.stringCoder(str);
+        int freeBytes = strlen + 2;
+        if (coder == 0) {
+            // If it is too long, it may be slow due to cache misses.
+            if (strlen < 256) {
+                countGreaterThanZero = JLA.countGreaterThanZero(str);
+            }
+            freeBytes += strlen - countGreaterThanZero; // 2 bytes
+        } else {
+            freeBytes += (strlen << 1); // 3 bytes
+        }
+        reserveSpace(freeBytes);
+
+        int start = this.offset;
+        int offset = start + 2;
+        byte[] elems = this.elems;
+
+        str.getBytes(0, countGreaterThanZero, elems, offset);
+        offset += countGreaterThanZero;
+
+        for (int i = countGreaterThanZero; i < strlen; ++i) {
+            char c = str.charAt(i);
+            if (c >= '\001' && c <= '\177') {
+                elems[offset++] = (byte) c;
+            } else if (c > '\u07FF') {
+                elems[offset    ] = (byte) (0xE0 | c >> 12 & 0xF);
+                elems[offset + 1] = (byte) (0x80 | c >> 6 & 0x3F);
+                elems[offset + 2] = (byte) (0x80 | c      & 0x3F);
+                offset += 3;
+            } else {
+                elems[offset    ] = (byte) (0xC0 | c >> 6 & 0x1F);
+                elems[offset + 1] = (byte) (0x80 | c      & 0x3F);
+                offset += 2;
+            }
+        }
+        int utflen = offset - start - 2;
+        if (utflen > 65535) {
```

Comment:
Previously we'd count bytes and throw before writing to the byte array. Perhaps this is still preferable, even if it comes at a cost.